internal/dedupe/dedupe.go

```diff
@@ -2,12 +2,22 @@
 package dedupe
 
 import (
+	"context"
 	"fmt"
-
-	"github.com/cloudevents/sdk-go/v2/event"
 )
 
-// GetEventKey creates a unique key from an event.
-func GetEventKey(namespace string, ev event.Event) string {
-	return fmt.Sprintf("%s-%s-%s", namespace, ev.Source(), ev.ID())
+// Deduplicator checks if an event is unique.
+type Deduplicator interface {
+	IsUnique(ctx context.Context, item Item) (bool, error)
+	Set(ctx context.Context, events ...Item) error
```

Comment:
Doesn't this change introduce a race condition? (Setting and checking in different steps)

Reply:
Depends on the usage. In the sink worker, it is okay because we parallelize processing per partition, and processing per partition is sequential while we also dedupe in the same batch in-memory.  

Inconsistency can happen in some edge cases still tough as we partition based on `subject`, so if someone sends in the same ID for a different subject at the exact same time, it can go to a parallel worker. It is an edge case as you shouldn't change the subject for the same ID. Optionally we can change the Kafka key (used by partitioning) to ID+Subject to protect against it. Originally we chose the subject as we needed data co-location with KSQLDB.